src/types/widget.ts

```diff
@@ -6,18 +6,55 @@ import {
   SearchParameters,
 } from './instantsearch';
 
-interface WidgetArgs {
-  helper: Helper;
+interface InitOptions {
+  instantSearchInstance: InstantSearch;
   state?: HelperState;
-  results?: SearchResults;
+  helper?: Helper;
+  templatesConfig?: object;
+  createURL?(state: HelperState): string;
+}
+
+interface RenderOptions {
   instantSearchInstance: InstantSearch;
+  templatesConfig?: object;
+  results?: SearchResults;
+  state?: HelperState;
+  helper?: Helper;
+  searchMetadata?: {
+    isSearchStalled: boolean;
+  };
+  createURL?(state: HelperState): string;
+}
+
+interface DisposeOptions {
+  helper?: Helper;
+  state: HelperState;
 }
 
+type UiState = {
+  [stateKey: string]: any;
+};
+
 export interface Widget {
-  init(options: WidgetArgs): void;
-  render(options: WidgetArgs): void;
-  dispose(options: any): SearchParameters;
-  getConfiguration?(previousConfiguration: SearchParameters): SearchParameters;
+  init?(options: InitOptions): void;
```

Comment:
This definition is not entirely true because a widget must at least have an `init` or a `render` method, or both ([see reference](https://www.algolia.com/doc/guides/building-search-ui/widgets/create-your-own-widgets/js#the-widget-lifecycle-and-api)).

This is however the easiest we can do right now, because it becomes complex to type otherwise.